_LowLevelCode/CPP/cpp_communicator/MPI_wrapper.h

```diff
@@ -1,25 +1,102 @@
 #include <vector>
+#include <list>
 #include <cmath>
 #include <mpi.h>
+#include "input_parser.h"
+
+/** Helper class to keep information on send message unit MPI framework reports delivered.
+*
+* in test mode also used to simulate send/receive operations.
+*/
+class SendMessHolder {
+public:
+    // Holder for ISend request property
+    MPI_Request theRequest;
+    // The holder for the tag of the message to send
+    int mess_tag;
+    // The holder for the address of the message to send
+    int destination;
+    // vector of the message contents, used as the buffer of the message contents until the message is received
+    std::vector<uint8_t> mess_body;
+      //SendMessHolder(SendMessHolder&& other) noexcept;
+    SendMessHolder() :
+        mess_tag(-1), destination(-1) {
+        this->theRequest = (MPI_Request)(-1);
+    }
+    SendMessHolder(uint8_t* pBuffer, size_t n_bytes, int dest_address, int data_tag);
+
+    void init(uint8_t* pBuffer, size_t n_bytes, int dest_address, int data_tag);
+};
 
 /* The class which describes a block of information necessary to process block of pixels */
 class MPI_wrapper {
 public:
 
-	MPI_wrapper():
-        labIndex(-1), numProcs(0){}
-	int init();
-	void close();
+    MPI_wrapper() :
+        labIndex(-1), numProcs(0), isTested(false),
+        async_queue_max_len_(10) {}
+    int init(bool isTested = false, int assynch_queue_max_len = 10, int data_mess_tag=5);
+    void close();
     void barrier();
+    void clearAll();
+    void labSend(int data_address, int data_tag, bool is_synchroneous, uint8_t* data_buffer, size_t nbytes_to_transfer);
+    void labProbe(const std::vector<int32_t> &data_address, const std::vector<int32_t> &data_tag, 
+        std::vector<int32_t> & addres_present, std::vector<int32_t> & tag_present);
+    void labReceive(int source_address, int source_data_tag, bool isSynchronous, mxArray* plhs[], int nlhs);
+    ~MPI_wrapper() {
+        this->close();
+    }
+    // index of the current MPI lab (worker)
+    int labIndex;
+    // total  number of MPI labs (workers)
+    int numProcs;
+    // test mode used to run various test operations over MPI_wrapper in single process, 
+    // when no real mpi exchange is initiated.
+    bool isTested;
+    // return the number of asynchroneous messages in the queue
+    size_t async_queue_len() {
+        return this->asyncMessList.size();
+    }
+    // the tag of message, containing data (processed differently, not yet implemented.)
+    static int data_mess_tag;
 
-	~MPI_wrapper() {
-		this->close();
-	}
+    //----------------------------------------------------------------------------------
+    // The methods used in unit tests -- have no meaning in real communications
 
-	int labIndex;  // index of the current MPI lab (worker)
-	int numProcs; // total  number of MPI labs (workers)
+    static bool MPI_wrapper_gtested;
+    // get access to the asynchroneous messages queue
+    std::list<SendMessHolder>* get_async_queue() {
```

Comment:
Is there a reason we're returning a poiner to the `asyncMessList` object here and not the object itself? i.e. 

```cpp
const inline std::list<SendMessHolder>& get_async_queue() {
    return this->asyncMessList
}
```
We should `const` wherever possible, and `inline` our getters (tells the compiler to simply put the returned object in place, skipping the function call).

Same thing for `get_sync_queue` below.

Reply:
because this may cause object copying and duplication I want to change object in-place. 
Accepting reference would to that, but then the receiver should always be with & and I think its more complicated.

Reply:
const pointer would probably be ok, but as its for tests, I wanted maximal flexibility (not used it but you never know in advance)

Reply:
But the `&` in the function type definition means there will be no copying, you're giving access to the unerlying object https://stackoverflow.com/questions/20091046/what-should-a-c-getter-return. 

In any case, if we're going to use pointers we should be using smart-pointers or we're going to run into memory leaks https://en.cppreference.com/book/intro/smart_pointers.

Reply:
Let's move discussion on code style into #36 where one bug in this code (do not understand why it was the bug but testing clearly've showed it) was fixed.

Pointers are generally less safe, and proven to be a problem while keeping an allocated memory, but work  well for referring to existing objects.

Reply:
very interesing discussion, but so petty when so much other work is wating attention....

Reply:
here I was thinking of using the same pointer to switch from synchronous to asynchronous containters. This is possible and should be used in tests only. 
access to the underlying object is what is intended....

Reply:
As I've already said, and just tested. whatever blog is telling
`    auto queue = wrap.get_async_queue();
    queue->rbegin()->theRequest = (MPI_Request)1;
`
makes and modifies the copy of the internal object if
`    std::list<SendMessHolder>& get_async_queue() {
        return this->asyncMessList;
    }
`
you need `auto &` to modify object and this is ugly.